here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/AbstractTask.java

```diff
@@ -398,63 +402,61 @@ protected final void unlock() {
-        assert threadCount >= 0L;
-        String author = context.getAuthor();
-        if (author != null) {
-          AtomicInteger authorUsage = AbstractTask.authorUsageMap.getOrDefault(author, new AtomicInteger(0));
-          int authorLimit = (int) (LIMIT * authorThreshold);
-          if (!internal && authorUsage.get() >= authorLimit) {
-            log.info(
-                "NAKSHA_ERR_REQ_LIMIT_4_PRINCIPAL - [Request Limit breached for Principal => appId,author,limit,crtValue] - ReqLimitForPrincipal {} {} {} {}",
-                context.getAppId(),
-                author,
-                authorLimit,
-                authorUsage);
-            throw new TooManyTasks(0, (long) (authorLimit), false);
-          }
-        }
-        if (!internal && threadCount >= LIMIT) {
+      long myIncrementedValue = AbstractTask.threadCount.updateAndGet(l -> {
+        if (!internal && AbstractTask.threadCount.get() >= LIMIT) {
           log.info(
               "NAKSHA_ERR_REQ_LIMIT_4_INSTANCE - [Request Limit breached for Instance => appId,author,limit,crtValue] - ReqLimitForInstance {} {} {} {}",
               context.getAppId(),
-              author,
+              getAuthorName(),
               LIMIT,
               threadCount);
-          throw new TooManyTasks(LIMIT, 0, true);
+          String errorMessage = "Maximum number of concurrent tasks reached for instance (" + LIMIT + ")";
+          throw new TooManyTasks(errorMessage);
         }
-        if (AbstractTask.threadCount.compareAndSet(threadCount, threadCount + 1)) {
-          if (author != null) {
-            AbstractTask.incrementAuthorUsage(author);
-          }
-          try {
-            state.set(State.START);
-            final Future<RESULT> future = threadPool.submit(this::init_and_execute);
-            return future;
-          } catch (RejectedExecutionException e) {
-            throw new TooManyTasks();
-          } catch (Throwable t) {
-            AbstractTask.threadCount.decrementAndGet();
-            if (author != null) {
-              AbstractTask.decrementAuthorUsage(author);
-            }
-            log.atError()
-                .setMessage("Unexpected exception while trying to fork a new thread")
-                .setCause(t)
-                .log();
-            throw new RuntimeException("Internal error while forking new worker thread", t);
-          }
-        }
-        // Conflict, two threads concurrently try to fork.
-      } while (true);
+        return l + 1;
+      });
+      assert myIncrementedValue >= 0L;
+      String principal = getPrincipalName();
+      if (principal != null) {
+        AtomicInteger authorUsageValue = principalUsageMap.merge(
+            principal, new AtomicInteger(0), (oldAuthorUsageValue, newAuthorUsageValue) -> {
+              int authorLimit = (int) (LIMIT * principalThreshold);
+              if (!internal && oldAuthorUsageValue.get() >= authorLimit) {
```

Comment:
pls use same convention everywhere. Our limits is based on principal and not author. so, avoid using "author" everywhere

Reply:
I totally agree to this one, clarify terminology.

IMHO we should extend the NakshaContext with a function getPrincipalId(), which should return the "author", when there is any, otherwise it should return the "appId". The reason is simple: When a user does use an application, we want to restrict this person to not overrun the service, therefore, we do not care what application he uses, the limit is bound to the person. However, if an application is run standalone without an authenticated user (e.g. a bot), we want to limit this application.

Apart of this, we need to consider that not all principals may have the same limit, so the limit should be considered per principal. Therefore, we should create a static atomic reference into which the application can load an interface that has one method being `getPrincipalLimit(context: NakshaContext): Int`, for example `static AtomicReference<IPrincipalManager> principalManager = new AtomicReference( new DefaultImpl())`. The default implementation can be simply, to calculate a limit for the machine for all principals and always to return the same value, but it allows us later to intercept certain principals.

Reply:
We decided to use Actor and new classes are required as mentioned.